godnsbl.go

```diff
@@ -97,91 +94,88 @@ func Blacklists() []string {
 	Listed bool `json:"listed"`
-	// RBL lists sometimes add extra information as a TXT record
-	// if any info is present, it will be stored here.
+	// RBL lists sometimes add extra information as a TXT record.
+	// If any info is present, it will be stored here.
 	Text string `json:"text"`
-	// Error represents any error that was encountered (DNS timeout, host not
-	// found, etc.) if any
-	Error bool `json:"error"`
-	// ErrorType is the type of error encountered if any
-	ErrorType error `json:"error_type"`
 }
 
 // Reverse the octets of a given IPv4 address
-// 64.233.171.108 becomes 108.171.233.64.
-func Reverse(ip net.IP) string {
-	if ip.To4() == nil {
-		return ""
-	}
-
-	splitAddress := strings.Split(ip.String(), ".")
-
-	for i, j := 0, len(splitAddress)-1; i < len(splitAddress)/2; i, j = i+1, j-1 {
-		splitAddress[i], splitAddress[j] = splitAddress[j], splitAddress[i]
+// '64.233.171.108' becomes '108.171.233.64'.
+// If it's not an v4 IP, returns nil.
+func Reverse(ip net.IP) net.IP {
+	ip4 := ip.To4()
+	if ip4 == nil {
+		return nil
 	}
 
-	return strings.Join(splitAddress, ".")
+	return net.IPv4(ip4[3], ip4[2], ip4[1], ip4[0])
 }
 
-func query(rbl string, host string, r *Result) {
-	r.Listed = false
-
-	lookup := fmt.Sprintf("%s.%s", host, rbl)
+func query(rbl string, ip net.IP) (*Result, error) {
+	lookup := fmt.Sprintf("%s.%s", Reverse(ip).String(), rbl)
 
 	res, err := net.LookupHost(lookup)
+	if err != nil && !recordNotFound(err) { // If the record doesn't exists, the IP isn't blacklisted
+		return nil, err
+	}
+
+	listed := false
+	text := ""
+
 	if len(res) > 0 {
-		r.Listed = true
-		txt, _ := net.LookupTXT(lookup)
+		listed = true
+
+		txt, err := net.LookupTXT(lookup)
+		if err != nil && !isNonFatalDNSError(err) {
+			return nil, fmt.Errorf("looking up TXT record: %w", err)
+		}
 
 		if len(txt) > 0 {
-			r.Text = txt[0]
+			text = txt[0]
 		}
 	}
 
-	if err != nil {
-		r.Error = true
-		r.ErrorType = err
-	}
+	return &Result{
+		Listed: listed,
+		Text:   text,
+	}, nil
 }
 
-// Lookup performs the search and returns the RBLResults.
-func Lookup(rblList string, targetHost string) (r RBLResults) {
-	r.List = rblList
-	r.Host = targetHost
+func recordNotFound(err error) bool {
+	var dnsErr *net.DNSError
+	if errors.As(err, &dnsErr) {
+		return dnsErr.IsNotFound
+	}
+
+	return false
+}
 
-	if ip, err := net.LookupIP(targetHost); err == nil {
-		for _, addr := range ip {
-			if addr.To4() != nil {
-				res := Result{}
-				res.Address = addr.String()
+// false means the error is either a fatal DNS error, or not even a DNS error.
+func isNonFatalDNSError(err error) bool {
+	var dnsErr *net.DNSError
+	if errors.As(err, &dnsErr) {
+		if dnsErr.Temporary() || dnsErr.IsNotFound {
+			return true
+		}
+	}
 
-				addr := Reverse(addr)
+	return false
+}
 
-				query(rblList, addr, &res)
+// Lookup performs the search and returns the RBLResults.
+func Lookup(rblList string, rawIP string) (*Result, error) {
```

Comment:
I removed the option to pass a domain name here.
It significantly simplifies things.
If you want to look up all addresses for a domain, you can just do the address lookups yourself 🤷